### DISCUSSION TITLE
how to do nested "with items" in orquesta?

### DISCUSSION THREAD

**Discussion by clindner85** [StackStorm/st2#5718] [Workflows] [answered]
Hi all,

During execution of a workflow, I want to find out if another workflow is already running (avoid duplicates).
I can only find out if it is a duplicate by checking one of the child workflows.

Basically, I need to process the result from st2.executions.list action=my.workflow status=running
That may give me one or many executions where each execution has several "children" exec IDs listed.

In python, I would do something like:
```
"""Pseudocode"""
executions = st2.executions.list(action="my.workflow", status="running")
for parent in executions:
    for child in parent.children:
        # do something
```

How can I do such a nested for loop in orquesta using "with items"? 

Would it be more effective and less complex to create a custom action and deal with this in python code?
Many thanks in advance!
Chris

**Comment by nzlosh** [comment, marked as answer]
To control workflow concurrency use Policies https://docs.stackstorm.com/reference/policies.html, since attempting to perform concurrency management from an action that checks St2 execution state will more than likely encounter unresolvable race conditions.

The is no concept of nesting `with-items` as per https://docs.stackstorm.com/orquesta/languages/orquesta.html#with-items-model, it's important to understand that results are not available per item iteration, but at the end of the `with-item` action in the form of a result list.  So `with-items` behaves nothing like a python loop.

If the StackStorm concurrency model doesn't fit your needs, you may have to look at an external distributed locking system, be it your own or a third parties.

**Reply by clindner85** [reply]
Many thanks for the fast reply!
I'll check the policies and probably redesign my workflow accordingly.

**Reply by bertyah** [reply]
To piggy back on this response, is concurrency Policies something that would work for Rate limiting actions that reach out to APIs?
